handler/oauth2/flow_authorize_jwt_token.go

```diff
@@ -0,0 +1,330 @@
+	}
+
+	// if client is authenticated, check grant types
+	if !c.CanSkipClientAuth(request) && !request.GetClient().GetGrantTypes().Has(grantTypeJwtBearer) {
+		return errorsx.WithStack(fosite.ErrUnauthorizedClient.WithHintf("The OAuth 2.0 Client is not allowed to use authorization grant \"%s\".", grantTypeJwtBearer))
+	}
+
+	return c.IssueAccessToken(ctx, request, response)
+}
+
+func (c *AuthorizeJwtGrantHandler) CanSkipClientAuth(requester fosite.AccessRequester) bool {
+	return c.SkipClientAuth
+}
+
+func (c *AuthorizeJwtGrantHandler) CanHandleTokenEndpointRequest(requester fosite.AccessRequester) bool {
+	// grant_type REQUIRED.
+	// Value MUST be set to "authorization_code"
+	return requester.GetGrantTypes().ExactOne(grantTypeJwtBearer)
+}
+
+func (c *AuthorizeJwtGrantHandler) validateTokenPreRequisites(token *jwt.JSONWebToken) error {
+	unverifiedClaims := jwt.Claims{}
+	if err := token.UnsafeClaimsWithoutVerification(&unverifiedClaims); err != nil {
+		return errorsx.WithStack(fosite.ErrInvalidGrant.
+			WithHint("Looks like there are no claims in JWT in \"assertion\" request parameter.").
+			WithWrap(err).WithDebug(err.Error()),
+		)
+	}
+	if unverifiedClaims.Issuer == "" {
+		return errorsx.WithStack(fosite.ErrInvalidGrant.
+			WithHint("The JWT in \"assertion\" request parameter MUST contain an \"iss\" (issuer) claim."),
+		)
+	}
+	if unverifiedClaims.Subject == "" {
+		return errorsx.WithStack(fosite.ErrInvalidGrant.
+			WithHint("The JWT in \"assertion\" request parameter MUST contain a \"sub\" (subject) claim."),
+		)
+	}
+
+	return nil
+}
+
+func (c *AuthorizeJwtGrantHandler) findPublicKeyForToken(ctx context.Context, token *jwt.JSONWebToken) (*jose.JSONWebKey, error) {
+	unverifiedClaims := jwt.Claims{}
+	if err := token.UnsafeClaimsWithoutVerification(&unverifiedClaims); err != nil {
+		return nil, errorsx.WithStack(fosite.ErrInvalidRequest.WithWrap(err).WithDebug(err.Error()))
+	}
+
+	var keyID string
+	for _, header := range token.Headers {
+		if header.KeyID != "" {
+			keyID = header.KeyID
+			break
+		}
+	}
+
+	keyNotFoundMsg := fmt.Sprintf(
+		"No public JWK was registered for issuer \"%s\" and subject \"%s\", and public key is required to check signature of JWT in \"assertion\" request parameter.",
+		unverifiedClaims.Issuer,
+		unverifiedClaims.Subject,
+	)
+	if keyID != "" {
+		key, err := c.AuthorizeJwtGrantStorage.GetPublicKey(ctx, unverifiedClaims.Issuer, unverifiedClaims.Subject, keyID)
+		if err != nil {
+			return nil, errorsx.WithStack(fosite.ErrInvalidGrant.WithHint(keyNotFoundMsg).WithWrap(err).WithDebug(err.Error()))
+		}
+		return key, nil
+	}
+
+	keys, err := c.AuthorizeJwtGrantStorage.GetPublicKeys(ctx, unverifiedClaims.Issuer, unverifiedClaims.Subject)
+	if err != nil {
+		return nil, errorsx.WithStack(fosite.ErrInvalidGrant.WithHint(keyNotFoundMsg).WithWrap(err).WithDebug(err.Error()))
+	}
+
+	claims := jwt.Claims{}
+	for _, key := range keys.Keys {
+		err := token.Claims(key, &claims)
```

Comment:
Given that we are validating the signature here and in `validateToken` as well I am wondering if we could do the (compute intense because probably RSA) computation just once instead?